File: tools/src/main/scala/com/threerings/nexus/streamergen/Processor.scala
//
// $Id$

package com.threerings.nexus.streamergen

import scala.collection.JavaConversions._

import java.util.Set

import javax.annotation.processing.{AbstractProcessor, Filer, ProcessingEnvironment}
import javax.annotation.processing.{RoundEnvironment, SupportedAnnotationTypes}
import javax.lang.model.SourceVersion
import javax.lang.model.element.{Element, TypeElement, Name}
import javax.tools.Diagnostic

/**
 * Generates {@code Streamer} implementations for {@link Streamable} classes.
 */
@SupportedAnnotationTypes(Array("*"))
class Processor extends AbstractProcessor {
  override def init (procenv :ProcessingEnvironment) {
    super.init(procenv)
    _filer = procenv.getFiler
    _scanner = new Scanner(procenv)
  }

  override def getSupportedSourceVersion :SourceVersion = SourceVersion.latest

  override def process (annotations :Set[_ <: TypeElement], roundEnv :RoundEnvironment) = {
    if (!roundEnv.processingOver) {
      for (elem <- roundEnv.getRootElements) elem match {
        case telem :TypeElement => {
          val metas = _scanner.scanUnit(elem)
          if (!metas.isEmpty) generate(telem, metas)
        }
        case _ => processingEnv.getMessager.printMessage(
          Diagnostic.Kind.WARNING, "Weird element? " + elem.getClass)
      }
    }
    false
  }

  protected def generate (elem :TypeElement, metas :Seq[ClassMetadata]) {
    Generator.generate(_filer, elem, metas)
  }

  protected var _scanner :Scanner = _
  protected var _filer :Filer = _
}

File: tools/src/main/java/com/threerings/nexus/streamergen/Processor.java
//
// $Id$

package com.threerings.nexus.streamergen;

import java.util.Set;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.annotation.processing.SupportedAnnotationTypes;

import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.TypeElement;

/**
 * Generates {@code Streamer} implementations for {@link Streamable} classes.
 */
@SupportedAnnotationTypes("*")
public class Processor extends AbstractProcessor
{
    @Override public void init (ProcessingEnvironment procenv) {
        super.init(procenv);
        _scanner = new Scanner(procenv);
    }

    @Override public SourceVersion getSupportedSourceVersion () {
        return SourceVersion.latest();
    }

    @Override public boolean process (Set<? extends TypeElement> annotations,
                                      RoundEnvironment roundEnv) {
        if (!roundEnv.processingOver()) {
            for (Element elem : roundEnv.getRootElements()) {
                _scanner.scanUnit(elem);
            }
        }
        return false;
    }

    protected Scanner _scanner;
}
